### PR TITLE
Fetched latest blocks from database

### LINKED ISSUE #55: Fetch latest blocks from database
# Fetch latest blocks from database

## ⚠️ Reading contributors guidelines to get assigned is MANDATORY!

[Read contributors guidelines](https://github.com/walnuthq/op-scan/issues/1)

## User stories

As a rollup developer using op-scan, I want to fetch the latest blocks data either from JSON-RPC (current behavior) or from a database fed from op-indexer.
The data source should be abstracted away in a dedicated function and is configured using an environment variable.

## Validation

It should fetch the exact same data whether calling the RPC directly or querying the database.
It MUST adhere to the existing design pattern already implemented in the tx details page.

## Implementation

Create a new helper function under `components/pages/blocks/fetch-blocks.ts`, use the same logic found in `components/pages/tx/fetch-transaction-details.ts`, a top level exported function which is data source agnostic and 2 separate functions fetching data from either JSON-RPC (this is already implemented here: https://github.com/walnuthq/op-scan/blob/main/src/lib/fetch-data.ts#L20-L28) or from the database (this is what has to be implemented).
Move the fetchLatestBlocks data fetching logic from the `lib/fetch-data.ts` to this helper function and just call this function in the blocks page top-level server component.
You will need to make sure the pagination logic is preserved when querying data via Prisma.

Run op-indexer as a background task in another terminal along the explorer to keep an up-to-date database mirroring what you'd fetch from the JSON-RPC endpoint.

## Resources

- fetch tx details helper function: https://github.com/walnuthq/op-scan/blob/main/src/components/pages/tx/fetch-transaction-details.ts
- data source agnostic function called in tx detail page: https://github.com/walnuthq/op-scan/blob/main/src/components/pages/tx/index.tsx#L10
- op-indexer getting started https://github.com/walnuthq/op-scan/blob/main/README.md#run-the-indexer

### PR DESCRIPTION
**Pull request by Ugo-X**
Resolves #55

### CODE REVIEW

File: src/components/pages/blocks/fetch-blocks.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    ),
+  );
+  return blocks.map(fromViemBlock);
+};
+
+const fetchLatestBlocksFromDatabase = async (
+  start: bigint,
+): Promise<Block[]> => {
+  const blocksPerPage = BigInt(process.env.NEXT_PUBLIC_BLOCKS_PER_PAGE);
+  const blocks = await prisma.block.findMany({
```
Comment (saimeunt) [inline src/components/pages/blocks/fetch-blocks.ts:24]: Please use [this pattern](https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries#select-specific-fields-of-included-relations) to also fetch the blocks transactions hashes, this is needed to display the correct number of transactions for each blocks.

File: src/components/pages/blocks/fetch-blocks.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+  return blocks.map((block) => ({
+    number: BigInt(block.number),
+    hash: block.hash as `0x${string}`,
+    parentHash: block.parentHash as `0x${string}`,
+    timestamp: BigInt(block.timestamp),
+    gasLimit: BigInt(block.gasLimit),
+    gasUsed: BigInt(block.gasUsed),
+    extraData: block.extraData as `0x${string}`,
+    transactions: [],
```
Comment (saimeunt) [inline src/components/pages/blocks/fetch-blocks.ts]: Use the block.transactions field here to construct an array of tx hashes.

**Review by saimeunt — CHANGES_REQUESTED** [review]
@Ugo-X This is almost correct, you just need to get the transactions count right.

**Comment by saimeunt** [comment]
@Ugo-X Also, please resolve the conflict.

File: src/components/pages/blocks/fetch-blocks.ts (hunk 2 above)
Comment (Ugo-X) [reply src/components/pages/blocks/fetch-blocks.ts]: done

File: src/components/pages/blocks/fetch-blocks.ts (hunk 1 above)
Comment (Ugo-X) [reply src/components/pages/blocks/fetch-blocks.ts:24]: i stil used the findMany but now ive fixed it and it shows the correct number of transactions for each block

**Comment by Ugo-X** [comment]
i increased the start value of the fetch latest block parameter as block 0 has 0 transactions

**Review by saimeunt — APPROVED** [review]

**Pull request merged**